lib/posix-process/signal/deliver.c

```diff
@@ -0,0 +1,355 @@
+
+	if (ks->ks_flags & SA_ONSTACK) {
+		UK_ASSERT(altstack->ss_flags & SS_ONSTACK);
+		UK_ASSERT(!(altstack->ss_flags & SS_DISABLE));
+		altstack->ss_flags &= ~SS_ONSTACK;
+	}
+
+	uk_mutex_unlock(&this_process->signal->lock);
+}
+
+/* Deliver a signal to a thread. The caller must check that the signal
+ * is not masked by the handling thread or ignored by the process or its
+ * default disposition.
+ */
+static int do_deliver(struct posix_thread *pthread, struct uk_signal *sig,
+		      struct uk_syscall_ctx *usc)
+{
+	struct posix_process *pproc;
+	uk_sigset_t saved_mask;
+	int signum;
+
+	UK_ASSERT(sig && sig->siginfo.si_signo);
+	UK_ASSERT(pthread);
+
+	pproc = tid2pprocess(pthread->tid);
+	UK_ASSERT(pproc);
+
+	signum = sig->siginfo.si_signo;
+
+	/* Execute the default action if a signal handler is not defined */
+	if (pproc->signal->sigaction[signum].ks_handler == SIG_DFL) {
+		/* Standard signals: Invoke signal-specific default action */
+		if (signum < SIGRTMIN) {
+			if (UK_BIT(signum) & SIGACT_CORE_MASK)
+				uk_sigact_core(signum);
+			else if (UK_BIT(signum) & SIGACT_TERM_MASK)
+				uk_sigact_term(signum);
+			else if (UK_BIT(signum) & SIGACT_STOP_MASK)
+				uk_sigact_stop(signum);
+			else if (UK_BIT(signum) & SIGACT_CONT_MASK)
+				uk_sigact_cont(signum);
+			else if (UK_BIT(signum) & SIGACT_IGN_MASK)
+				uk_sigact_ign(signum);
+			else
+				UK_BUG();
+		} else {
+			/* Real-time signals: SIG_TERM by default */
+			uk_sigact_term(signum);
+		}
+		return 0;
+	}
+
+	uk_mutex_lock(&pproc->signal->lock);
+
+	/* Save original mask and apply mask from sigaction */
+	saved_mask = pthread->signal->mask;
+	uk_sigorset(&pthread->signal->mask , &pproc->signal->sigaction[signum].ks_mask);
+
+	/* Also add this signal to masked signals */
+	if (!(pproc->signal->sigaction[signum].ks_flags & SA_NODEFER))
+		SET_MASKED(pthread, signum);
+
+	uk_mutex_unlock(&pproc->signal->lock);
+
+	/* Execute handler */
+	handle_self(sig, &pproc->signal->sigaction[signum], usc);
+
+	uk_mutex_lock(&pproc->signal->lock);
+
+	/* Restore original mask */
+	pthread->signal->mask = saved_mask;
+
+	/* If SA_RESETHAND flag is set, restore the default handler */
+	if (pproc->signal->sigaction[signum].ks_flags & SA_RESETHAND)
+		pprocess_signal_sigaction_clear(&pproc->signal->sigaction[signum]);
+
+	uk_mutex_unlock(&pproc->signal->lock);
+
+	return 0;
+}
+
+/* Deliver pending signals of a given process. We deliver each
+ * pengind signal to the first thread that doesn't mask that
+ * signal.
+ */
+static int deliver_pending_proc(struct posix_process *proc,
+				struct uk_syscall_ctx *usc)
+{
+	struct posix_thread *thread, *threadn;
+	struct posix_thread *this_thread;
+	struct uk_signal *sig;
+	int handled_cnt = 0;
+	bool handled;
+
+	UK_ASSERT(proc);
+	UK_ASSERT(usc);
+
+	this_thread = tid2pthread(uk_gettid());
+
+	for (int i = 0; i < SIG_ARRAY_COUNT; i++) {
+		handled = false;
+
+		/* Skip if the process ignores this signal altogether */
+		if (IS_IGNORED(proc, i))
```

Comment:
Would it help to add an assert here that the ignored signal is not `SIGSTOP`/`SIGCONT`?

Reply:
Can you please elaborate on this one?